schema.cc

```diff
@@ -102,21 +102,30 @@ std::ostream& operator<<(std::ostream& os, ordinal_column_id id)
     return os << static_cast<column_count_type>(id);
 }
 
-std::unique_ptr<dht::i_partitioner> default_partitioner;
+thread_local std::map<std::tuple<sstring, unsigned, unsigned>, std::unique_ptr<dht::i_partitioner>> cached_partitioners;
+sstring default_partitioner_name = "org.apache.cassandra.dht.Murmur3Partitioner";
+unsigned default_partitioner_ignore_msb = 12;
 
 void schema::set_default_partitioner(const sstring& class_name, unsigned ignore_msb)
 {
-    default_partitioner = dht::make_partitioner(class_name, smp::count, ignore_msb);
+    default_partitioner_name = class_name;
+    default_partitioner_ignore_msb = ignore_msb;
+}
+
+static const dht::i_partitioner& get_partitioner(const sstring& name, unsigned shard_count, unsigned ignore_msb) {
+    auto it = cached_partitioners.find({name, shard_count, ignore_msb});
+    if (it == cached_partitioners.end()) {
+        auto p = dht::make_partitioner(name, shard_count, ignore_msb);
```

Comment:
This isn't really a cache, but interning. Items from the cache are never invalidated and reclaimed. 

That's not a problem since we have O(1) partitioners.

How does the partitioner interact with frozen_schema? I missed where we freeze and re-create it.

Reply:
I renamed the variable to partitioners.

frozen_schema works in terms of schema_mutation which at the moment ignores partitioner. This is fine for now because the only custom partitioner will be for CDC and it will be always set to the same partitioner. Later when we will be making it possible to set any partitioner on any table we will have to include partitioner field in both schema_mutation and frozen_schema.

Reply:
Not sure this is correct. global_schema_ptr uses frozen_schema, and in turn is used when queries or writes cross shards. So if someone queries from the cdc table from the wrong shard, we lose the partitioner.

Reply:
You're right. It seems frozen_schema will have to handle partitioner as well.

Reply:
What's the conclusion of this discussion thread? Did you fix it (in this case please resolve the issue)? Left a FIXME and/or a issue? Or what?

Reply:
This is fixed in this PR.